runtime/src/bank.rs

```diff
@@ -5330,11 +5304,76 @@ impl Bank {
         );
+    }
+
+    /// load accounts with pubkeys in 'partition'
+    /// collect rent
+    /// store accounts, whether rent was collected or not
+    /// update bank's rewrites set for all rewrites that were skipped
+    /// if 'just_rewrites', function will only update bank's rewrites set and not actually store any accounts
+    /// return # accounts loaded
+    fn collect_rent_in_partition(
+        &self,
+        partition: Partition,
+        just_rewrites: bool,
+        metrics: &RentMetrics,
+    ) -> usize {
+        let subrange_full = Self::pubkey_range_from_partition(partition);
+
+        let mut hold_range = Measure::start("hold_range");
+        let thread_pool = &self.rc.accounts.accounts_db.thread_pool;
+        self.rc
+            .accounts
+            .hold_range_in_memory(&subrange_full, true, thread_pool);
+        hold_range.stop();
+        metrics.hold_range_us.fetch_add(hold_range.as_us(), Relaxed);
+
+        // divide the range into num_threads smaller ranges and process in parallel
+        // Note that 'pubkey_range_from_partition' cannot easily be re-used here to break the range smaller.
+        // It has special handling of 0..0 and partition_count changes affect all ranges unevenly.
+        let num_threads = crate::accounts_db::quarter_thread_count() as u64;
+        let sz = std::mem::size_of::<u64>();
+        let start_prefix = Self::prefix_from_pubkey(subrange_full.start());
+        let end_prefix_inclusive = Self::prefix_from_pubkey(subrange_full.end());
+        let range = end_prefix_inclusive - start_prefix;
+        let increment = range / num_threads;
+        let account_count = AtomicUsize::default();
+        thread_pool.install(|| {
+            (0..num_threads).into_par_iter().for_each(|chunk| {
+                let offset = |chunk| start_prefix + chunk * increment;
+                let start = offset(chunk);
+                let last = chunk == num_threads - 1;
+                let merge_prefix = |prefix: u64, mut bound: Pubkey| {
+                    bound.as_mut()[0..sz].copy_from_slice(&prefix.to_be_bytes());
+                    bound
+                };
+                let start = merge_prefix(start, *subrange_full.start());
+                let mut load = Measure::start("load");
+                let accounts = if last {
+                    let end = *subrange_full.end();
+                    let subrange = start..=end; // IN-clusive
+                    self.rc
+                        .accounts
+                        .load_to_collect_rent_eagerly(&self.ancestors, subrange)
+                } else {
+                    let end = merge_prefix(offset(chunk + 1), *subrange_full.start());
+                    let subrange = start..end; // EX-clusive, the next 'start' will be this same value
+                    self.rc
+                        .accounts
+                        .load_to_collect_rent_eagerly(&self.ancestors, subrange)
```

Comment:
nit: Looks like this is duplicated in the `if` case above and could be pulled out after the whole `if-else` block.

Reply:
One is an inclusive range, other is exclusive. So these are 2 different instances of the target function that takes a trait.